text/0486-deprecate-mouseenter.md

```diff
@@ -0,0 +1,137 @@
+and relying on the events bubbling up to that element (aka event delegation). There they 
+are processed and invoke any matching `Ember.Component` event handler method with 
+the same (camel-cased) method as the event type. Same for element-space `{{action}}` 
+modifiers.
+
+> Note: for a "Deep Dive on Ember Events" and how they differ from "native events" I refer to 
+Marie Chatfield's excellent 
+[blog post](https://medium.com/square-corner-blog/deep-dive-on-ember-events-cf684fd3b808)
+or [EmberConf talk](https://youtu.be/G9hXjjHFJVs)
+
+This works fine in general, but `mouseenter`/`mouseleave` events are special as they do
+not bubble. In the past it still worked nevertheless as jQuery transparently handled this
+for us, as it had special support for event delegation for these events, essentially by using 
+(bubbling) `mouseover` events to replicate the semantics of `mouseenter`/`mouseleave` events.
+
+When support for jQuery-less apps was introduced, this [left a hole](https://github.com/emberjs/ember.js/issues/16591)
+in the jQuery-less EventDispatcher implementation. But as support for those events was and
+still is part of Ember's pubic API, we had no chance other than to [implement support
+for jQuery-less apps](https://github.com/emberjs/ember.js/pull/16603) using the same 
+`mouseover` based approach.
+
+This however comes with a cost: besides an [unresolved issue](https://github.com/emberjs/ember.js/issues/17228)
+the implementation has some performance drawbacks, as it has to process every `mouseover` event on 
+*any* element, create fake `mouseenter`/`mouseleave` events and try to dispatch them, even when 
+not a single component/action needs them.  
+
+Deprecating support for `mousemove` is also proposed, which is a (bubbling) event that does not have the higher 
+implementation cost as `mouseenter`/`mouseleave`, but nevertheless requires the EventDispatcher to optimistically handle
+these extremely high-volume events.
+
+While efforts to make this more "pay as you go" are [possible](https://github.com/emberjs/ember.js/pull/17911), 
+the trade-off of keeping support around still seems unfavorable, as these events fire so
+frequently, while they are (most certainly) very rarely used.
+
+This is even more so given that Glimmer Components with their outerHTML semantics do not 
+work with event handler methods, and `{{action}}` will eventually fade away in favor of
+`{{on}}` using native `addListener()`.
+
+## Transition Path
+
+When a `mouseenter`/`mouseleave`/`mousemove` event is dispatched by the EventDispatcher to a component
+or `{{action}}` listening to that event type, a deprecation warning will be issued.
+
+The linked migration guide will cover these use cases:
+
+#### Ember.Component
+
+Before:
+
+```js
+import Component from '@ember/component';
+
+export default Component.extend({
+  mouseEnter(e) {
+    // do something
+  }
+});
+```
+
+After:
```

Comment:
This "after" example is pretty confusing to read/review. Here are a couple concrete things to be improved:

* The nuance that it is using `mouseenter` vs `mouseEnter` is **very** subtle. I'd like to see this tweaked to be something a bit less confusing (like `handleMouseEnter`) and use `@action` instead of manually binding in `didInsertElement`
* Would love to see two "After" examples, one similar to the existing one (with the tweaks suggested above) and anther for folks willing to migrate to "outer HTML" in their template (and then they could directly use `{{on` in the template, instead of `didInsertElement`).

Reply:
@rwjblue by using `@action` you imply changing that to a native class? I am fine with that, just thought that as the Guides are still showing "legacy" class examples, I should stick to them as well for now?

Reply:
@rwjblue updated the examples according to your feedback! Using native classes now.

Reply:
Sorry I missed this yesterday! You can use `action` in classic components:

```js
export default Component.extend({
  handleMouseEnter: action(function(event) {

  })
})
```

Reply:
Ah, makes sense. Though I haven't seen that yet in the wild...

@rwjblue do you want me to revert my changes back to classic components? 

I just wonder if we shouldn't stick with native classes in the examples, as native classes and decorators support have already landed, and `{{on}}` hasn't yet AFAIK. And given that it is going to take some time until the deprecation hits a stable version, we might already have landed Octane and native class examples in the Guides?

Reply:
Ya, this seems reasonable. It's an easy change to move back if we want to, and I don't think its a massive difference to "should we move forward".